src/shared/bootspec.c

```diff
@@ -755,45 +757,203 @@ static int boot_entry_load_unified(
-        if (ret_cmdline)
-                *ret_cmdline = TAKE_PTR(cmdline);
+        _cleanup_free_ IMAGE_SECTION_HEADER *sections = NULL;
+        _cleanup_free_ PeHeader *pe_header = NULL;
+        int r;
+
+        r = find_sections(fd, path, &sections, &pe_header);
+        if (r < 0)
+                return r;
+
+        r = find_cmdline_section(fd, path, sections, pe_header, ret_cmdline);
+        if (r < 0)
+                return r;
 
         return 0;
 }
 
+static int boot_entries_find_unified_addons(
+                BootConfig *config,
+                const char *root,
+                const char *addon_dir,
+                BootEntryAddon **addons) {
+
+        _cleanup_closedir_ DIR *d = NULL;
+        _cleanup_free_ char *full = NULL;
+        int r;
+
+        assert(addons);
+
+        r = chase_and_opendir(addon_dir, root, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, &full, &d);
+        if (r == -ENOENT)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to open '%s/%s': %m", root, addon_dir);
+
+        LIST_HEAD_INIT(*addons);
+        FOREACH_DIRENT(de, d, return log_error_errno(errno, "Failed to read %s: %m", full)) {
+                _cleanup_free_ char *j = NULL, *cmdline = NULL;
+                _cleanup_close_ int fd = -EBADF;
+                BootEntryAddon *addon = NULL;
+
+                if (!dirent_is_file(de))
+                        continue;
+
+                if (!endswith_no_case(de->d_name, ".addon.efi"))
+                        continue;
+
+                fd = openat(dirfd(d), de->d_name, O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOFOLLOW|O_NOCTTY);
+                if (fd < 0) {
+                        log_warning_errno(errno, "Failed to open %s/%s, ignoring: %m", full, de->d_name);
+                        continue;
+                }
+
+                r = config_check_inode_relevant_and_unseen(config, fd, de->d_name);
+                if (r < 0)
+                        return r;
+                if (r == 0) /* inode already seen or otherwise not relevant */
+                        continue;
+
+                j = path_join(full, de->d_name);
+                if (!j)
+                        return log_oom();
+
+                if (find_addon_sections(fd, j, &cmdline) < 0)
+                        continue;
+
+                addon = new(BootEntryAddon, 1);
+                if (!addon)
+                        return log_oom();
+
+                addon->location = strdup(j);
+                if (!addon->location)
+                        return log_oom();
+
+                addon->cmdline = TAKE_PTR(cmdline);
+                LIST_INIT(addon_list, addon);
+                LIST_APPEND(addon_list, *addons, addon);
+        }
+        return 0;
+}
+
+static int boot_entries_find_unified_global_addons(
+                BootConfig *config,
+                const char *root,
+                const char *d_name) {
+
+        assert(config);
```

Comment:
functions that do nothing else but wrap some other call we usually don't do any additional assert checking, if the inner call already does that anyway (here it doesn't, but it should, see above)